src/clients/FtpClient.cc

```diff
@@ -930,9 +952,16 @@ Ftp::Client::maybeReadVirginBody()
 
     debugs(9,5,"queueing read on FD " << data.conn->fd);
 
-    typedef CommCbMemFunT<Client, CommIoCbParams> Dialer;
-    entry->delayAwareRead(data.conn, data.readBuf->space(), read_sz,
-                          JobCallback(9, 5, Dialer, this, Ftp::Client::dataRead));
+    const auto amountToRead = entry->bytesWanted(Range<size_t>(0, read_sz));
+
+    if (amountToRead <= 0) {
+        delayRead();
+        return;
+    }
+
+    using ReadDialer = CommCbMemFunT<Client, CommIoCbParams>;
+    AsyncCall::Pointer readCallback = JobCallback(9, 5, ReadDialer, this, Client::dataRead);
+    comm_read(data.conn, data.readBuf->space(), amountToRead, readCallback);
```

Comment:
Please use the AsyncCall-aware Comm::Read / Comm::ReadNow API instead of the deprecated comm_read().

Usage of the new API is essentially what you are doing here ... except that this block of code should **only** pass the ```readCallback``` object to ```Comm::Read()```. The buffer calculations go in ```Client::dataRead()``` logic to setup for a ```Comm::ReadNow()``` fetching the data (no longer passed to the handler as a callback parameter).

Same change request for all the other AsyncJob methods equivalent to this delayAwareRead(). The gopher change and other old non-AsyncJob callbacks are fine doing this old way.

Reply:
I  started refactoring at first, addressing your request, but then realized that we cannot simply substitute the old comm_read(), requiring MemBuf with the new Comm::Read, requiring SBuf as an input buffer. Probably we could go further, converting DataChannel::readBuf to SBuf, but I am afraid this is a little out of the scope. There are other(minor) difficulties with this refactoring as well. For example, should we remove the deprecated replyBodySpace()? What to do with the strange 'read_sz<2' condition? Overall, I would do this refactoring in a separate PR, adding a TODO here instead.

> Same change request for all the other AsyncJob methods

I see only two comm_read() occurrences in this PR: this one and the Gopher's (which is out of the question).

Reply:
FWIW, I agree with Eduard that the requested change can and should be considered out of this PR scope. Properly upgrading FtpClient to use Comm::ReadNow() would require converting _both_ control and data channels to use SBuf storage and modern Comm reading APIs, triggering quite a bit of side effects that deserve careful analysis and review. All that work is useful but is best done in a dedicated PR, especially since this PR does not increase deprecated comm_read() usage but essentially "inlines" and "localizes" the old one (i.e. we can say that the PR is not making things worse in this regard).

Reply:
Okay, request withdrawn.